runtime/jvmti/jvmtiHelpers.c

```diff
@@ -108,18 +112,38 @@ getVMThread(J9VMThread * currentThread, jthread thread, J9VMThread ** vmThreadPt
 		}
 		return JVMTI_ERROR_INVALID_THREAD;
 	} else {
-		threadObject = *((j9object_t*) thread);
+		threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
 		if (currentThread->threadObject == threadObject) {
 			*vmThreadPtr = currentThread;
 			return JVMTI_ERROR_NONE;
 		}
 	}
 
-	/* Make sure the vmThread stays alive while it is being used */
-
+	/* Make sure the vmThread stays alive while it is being used. */
 	omrthread_monitor_enter(vm->vmThreadListMutex);
-	if (!J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) || ((targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject)) == NULL)) {
+#if JAVA_SPEC_VERSION >= 19
+	isVirtualThread = IS_VIRTUAL_THREAD(currentThread, threadObject);
+	if (isVirtualThread) {
+		/* Make sure the virtual thread stays alive while it is being used. */
+		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+		jint vthreadState = J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, threadObject);
+		j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(currentThread, threadObject);
+		targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, carrierThread);
+		isThreadAlive = (vthreadState != JVMTI_VTHREAD_STATE_NEW) && (vthreadState != JVMTI_VTHREAD_STATE_TERMINATED);
+	} else
+#endif /* JAVA_SPEC_VERSION >= 19 */
+	{
+		targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+		isThreadAlive = J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject);
+	}
+
+	if (!isThreadAlive || (targetThread == NULL)) {
```

Comment:
If the virtual thread is yielded, will the carrier thread be null and return not alive?
Also `NULL == targetThread`

Reply:
Yes. See [VirtualThread.java#L367-L370](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L367-L370). In [unmount](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L345), `VirtualThread.carrierThread` is set to `NULL` during yield. So, the `targetThread` will be `NULL`. Will fix `NULL == targetThread`. Also, a `NULL` check is missing for `VirtualThread.carrierThread`.

Reply:
Updated.

Reply:
If it's yielded, the targetThread will remain null and return "not alive" even though yielded is considered alive. Could it set it to currentThread if yielded? I need it for the tls get/set.

Reply:
Also do the jvmti functions need to have the carrier J9VMThread to work for virtual threads? Can we return vmThread = currentThread always?

Reply:
> Also do the jvmti functions need to have the carrier J9VMThread to work for virtual threads? Can we return vmThread = currentThread always?

It depends. Most of the time, the virtual thread's information will be derived from the carrier thread (J9VMThread) during inspection. The JVMTI functions won't behave correctly with `vmThread = currentThread`. For information which does not depend on the carrier thread, `vmThread = currentThread` can be used.

> If it's yielded, the targetThread will remain null and return "not alive" even though yielded is considered alive. Could it set it to currentThread if yielded? I need it for the tls get/set.

Some locations will break with `targetThread = currentThread (from getVMThread)` if a virtual thread is yielded. For tls get/set, we can make an exception for virtual threads if it is yielded. In general, we will have to evaluate each JVMTI function individually for such needs.

Reply:
Maybe we can return a different error if a virtual thread is yielded? Then handle accordingly based on the jvmti function.
Also this means that the `SetLocal(Object|Int|Long|Float|Double) ...` functions will not be fixed by this PR

Reply:
i can return the value of `isVirtual` from `getVMThread` via an input parameter; return `JVMTI_ERROR_NONE`; but, `targetThread` will be set to NULL. if `isVirtual && (JVMTI_ERROR_NONE == rc) && (NULL == targetThread)`, we can use the `currentThread` for evaluating the virtual thread detail. `currentThread` is derived in most cases.  for `SetLocal*`, the yielded virtual thread will need to be stack walked on the current thread. ++ @gacholio @fengxue-IS @tajila